### ISSUE TITLE
accuracy problem

### ISSUE THREAD

**Issue by bewilderss** [facebookresearch/videoseal#38]
Hi, I transformed specific string to tensor and embed to video. I tried random string and specific string, the accuracy for random string is 88.8% to 91.5%. the accuracy for specific string in different video is around 92%. Is there any method to improve my accuracy? The model I used is videoseal_1.0.

**Comment by pierrefdz** [comment]
Can you share with me what you ran?

If you're using `model.detect` could you try the `detect_and_aggregate` method instead?
See https://github.com/facebookresearch/videoseal/blob/main/docs/torchscript.md#video-watermarking
The `detect` computes the bit acc per frame, then you average and show the results. The `detect_and_aggregate` averages the soft message across all frames, then thresholds and get the overall message, which reduces bit error across frames.

**Comment by bewilderss** [comment]
Thank you, I tried you https://github.com/facebookresearch/videoseal/blob/main/docs/torchscript.md#video-watermarking, but the message acc rate is only around 50%. I don't know what goes wrong. 

Here is the code I ran which accuracy around 95%-100%, I want to know is there any method to keep the accuracy in 100%.

import torch
import torchvision
import videoseal
import cv2
import numpy as np
import time
from videoseal.evals.metrics import bit_accuracy


def str_to_binary_tensor(input_string):
    padded_string = input_string.ljust(12)
    binary_representation = ''.join(format(ord(char), '08b') for char in padded_string)
    binary_list = [int(bit) for bit in binary_representation]
    return torch.tensor([binary_list])


def binary_tensor_to_string(tensor):
    binary_str = ''.join(str(i) for i in tensor.flatten().tolist())
    result = ""
    for i in range(0, len(binary_str), 8):
        byte = binary_str[i:i + 8]
        char = chr(int(byte, 2))
        result += char
    return result


def embed_watermark(video_path, output_path, watermark_str, fps=25):
    model = videoseal.load("videoseal")
    message_tensor = str_to_binary_tensor(watermark_str)

    video, _, _ = torchvision.io.read_video(video_path, pts_unit='sec')
    video = video.permute(0, 3, 1, 2).float() / 255.0
    video = video[:fps * 3]

    start_time = time.time()
    outputs = model.embed(video, message_tensor, is_video=True)
    elapsed_time = time.time() - start_time
    print(f"embedding time: {elapsed_time:.4f} s")

    video_w = outputs["imgs_w"]
    video_w = (video_w * 255).byte().permute(0, 2, 3, 1)

    height, width, _ = video_w[0].shape
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    out = cv2.VideoWriter(output_path, fourcc, fps, (width, height))

    for frame in video_w:
        frame_bgr = cv2.cvtColor(frame.cpu().numpy(), cv2.COLOR_RGB2BGR)
        out.write(frame_bgr)
    out.release()
    print(f"✅ video is saving to {output_path}")

    return message_tensor


def extract_watermark(video_path):
    model = videoseal.load("videoseal")
    video, _, _ = torchvision.io.read_video(video_path, output_format="TCHW", pts_unit='sec')
    video = video.float() / 255.0

    start_time = time.time()
    msg_extracted = model.extract_message(video)
    elapsed_time = time.time() - start_time
    print(f"extracting time: {elapsed_time:.4f} s")

    return msg_extracted


def main():
    input_video = "./assets/videos/sap_3.mp4"
    output_video = "./assets/videos/sap_3_wm.mp4"
    watermark_string = "string"

    original_tensor = embed_watermark(input_video, output_video, watermark_string)
    extracted_tensor = extract_watermark(output_video)

    original_str = binary_tensor_to_string(original_tensor)
    extracted_str = binary_tensor_to_string(extracted_tensor)

    print(f"🎯 original string: {original_str}")
    print(f"🧾 extracting string: {extracted_str}")

    accuracy = bit_accuracy(original_tensor, extracted_tensor).nanmean().item()
    print(f"✅ Bit Accuracy: {accuracy:.3f}")


if __name__ == "__main__":
    main()

**Comment by pierrefdz** [comment]
Thanks, there were some errors in the torchscript script that should be fixed now.

Some accuracy is lost when saving the video because of the codec, it will be hard to always have 100% accuracy for these. What you can do:
1/ Change the step_size to 1 or 2 (model.step_size = 1). This will make embedding a bit slower, but should improve the final accuracy.
2/ Use error correcting codes on the message. It will restrict a bit the number of bits but make them more reliable. There's an example of such ECC here: https://github.com/tancik/StegaStamp/blob/master/decode_image.py

**Comment by pierrefdz** [comment]
In the demo we used a Hamming (15,11) code on the full message. Each character is encoded with 5 bits (which represent 2^5=32 characters).

Otherwise you can try something like the following, where we average the extracted soft message in groups of 2, 4, etc. bits  then threshold https://gist.github.com/pierrefdz/b71cabd0bc0a4c9a6a3a0f93e7ccd55e

**Comment by pierrefdz** [comment]
Hi, thanks!
Everything is licensed under MIT, see https://github.com/facebookresearch/videoseal/blob/main/LICENSE
